### DISCUSSION TITLE
Question about analytics collection

### DISCUSSION THREAD

**Discussion by AndrewFarley** [stepci/stepci#79] [Q&A] [answered]
### What happened?

So, a few thoughts...
* I believe you should be transparent about analytics on an open source project.  It should be on your main README.md and it should be very clear this library captures analytics.  It is at least decent of you to show about your privacy statement when running it via the CLI.  But, if this was used as a module and not as an executable, it wouldn't show.  It is decent of you to show it here: https://docs.stepci.com/legal/privacy.html though, it would be _more_ decent to show directly in the CLI how to disable it.  You need to be very clear and up front, in your README, about analytics and how to opt out I'd say.  It's one of those compliance/regulatory no-no's and just an general rude thing to do.
* I also believe that which isn't related to what the user wants to do, should NEVER interrupt, harass, or alarm the user or affect their experience with your tool negatively.  Unfortunately, the currently analytics implementation fails in this regards (see below).

So, I'm on an home network with PiHole blocking analytics traffic.  I was saddened to see this framework has analytics baked into it.  Also, quite alarmed of the time I run this framework I get bombarded with pages full of an exception call related to this failure of calling analytics which had nothing to do with me using this framework.

If you want to do **background** analytics on a framework like this, it needs to be **dead silent** unless the user specifies otherwise (eg: debug/verbose mode).  Or if not dead silent, at least very concise.

Currently it's pages of exception code when it fails which is kinda alarming which takes away from your core logic/code, as seen here: https://gist.github.com/376434c663b23b8e73767d2d748b9017

### What did you expect to happen?

1. The README.md should be updated and have working examples of how to disable the analytics code up-front, not shady links away.
1. The failure/exception regarding anything analytics related should be something decent, graceful, and concise such as the following unless the user specifies to be more verbose (eg: by enabling verbose/debug output mode)

```bash
$ npx stepci run workflow.yml

 PASS  example

Tests: 0 failed, 1 passed, 1 total
Steps: 0 failed, 0 skipped, 1 passed, 1 total
Time:  0.985s, estimated 1s

Workflow passed after 0.985s
Give us your feedback on https://step.ci/feedback
NOTE: Unable to report analytics, timeout
```

### Version

2.5.7

### Environment

v18.8.0

**Comment by AndrewFarley** [comment]
Also, wanted to say I do love this project and what you've done with it so far, so don't let my issue here take away from that.  I'm critical about things related to privacy and open-source code.  So let me just say thanks, love this project's foundation so far and your open-source contribution.

I'm considering engineering a cloud product for a client based on this.  :)  Let you know if/how and when I can open source it.  Cheers!

**Comment by mishushakov** [comment, marked as answer]
Hey,
Sorry it took a while to get back to you.

I wanted to thank you for the detailed write-up, I really appreciate that!

Here are my points:

1. We have now added a privacy section in the readme, this is not something we were trying to hide. We link the webpage in the CLI, because on the webpage you get more detailed information about what we collect, why and how you can disable it. Analytics in our project is optional. Besides that, we do not collect any analytics when it is used as a module (the runner doesn't include any analytics)
2. I understand the frustration and it must be an issue with PostHog, the analytics provider, not with our tool. I will forward it to them, but in the meanwhile you can just disable analytics by using the environment variable (`STEPCI_DISABLE_ANALYTICS`). The errors should disappear after that

Besides that, the project is open-source. If there's something you do not like about it, you can always fork the code-base and adjust it to your own needs

I'm very glad you're loving this project, if you want to build something on top of it, feel free to send me a mail to mish at stepci, I'd be glad to answer any technical questions

Thanks 🙏
